### ISSUE TITLE
Consider a more modern scripting language

### ISSUE THREAD

**Issue by francisdb** [vpinball/vpinball#568]
**Is your feature request related to a problem? Please describe.**

`VBScript` is ancient technology and hard to get 100% compatible cross-platform

**Describe the solution you'd like**

A more modern scripting language like `JS` that most devs/ppl these days have experience with. 

**Describe alternatives you've considered**

Fixing all existing and upcoming issues in wine vbscript. Might fix the `cross-platform` issues but will not improve the part where not many ppl know `vbs` these days.

**Additional context**

Of course this can't be a big bang migration but one needs to start somewhere so there needs to be some marker for JS scripts allowing to pick a different runtime.

**Comment by francisdb** [comment]
Looks like MS officially deprecated VBScript on October 2023

> VBScript is deprecated. In future releases of Windows, VBScript will be available as a feature on demand before its removal from the operating system. For more information, see Resources for deprecated features.     October 2023

https://learn.microsoft.com/en-us/windows/whats-new/deprecated-features

**Comment by thelayeredmind** [comment]
Hello, Just wanted to drop by and say what an amazing project this is. I rediscovered my old nostalgic passion for the game recently! So props to you guys. 

I would like to chime in into the discussion and agree that a creative developer myself (mainly experienced with Game Engines, modern Scripting languages like JS and Python or more exotic languages like Clojure or Haskell) I was also kind of shocked of the outdated language. @francisdb

Just some ideas, which might be completely out of scope or irrelevant (based on my lacking knoweldge of architecture and VSBasic). Is there any way to have a hybrid language ecosystem, where one language compiles or interfaces with the other? 

1. Compile / Transpile to VBScript: I have seen multiple examples of this throughout other projects, like in Neovim which uses Lua for scripting and configuration and the community has Fennel that adds a LISP flavour to it, compiling and integrating seamlessly into the system. Or languages like ClojureScript, Elm and PureScript compiling to javascript and integrating therefor functional languages seemlessly in web frontends.

2. .NET Integration: Does VBScript integrate with .NET similarly to Visual Basic, if so, maybe C# / F# could be an option?
Which might be a bit overkill for scripting a humble pinball table.

3. Have I mentioned Lua 😅 ? Seems to be a very popular language for scripting DSLs in many configurable interactive environments, and may even attract a wider programmer audience. I have also noticed an affinity for C/C++ projects when it comes to Lua scripting integration.

I hope I don't offend anybody with these suggestions. Just wanted to share my exitement for different options.
I would to give it a shot myself in the future, right now I am fumbling with my own share of projects and problems.

**Comment by francisdb** [comment]
Lua is indeed an other valid alternative.

I think @jsm174 has some experience transpiling vbscript to JS for [VisualPinball.Engine](https://github.com/freezy/VisualPinball.Engine)
Unfortunately the scripting is is not only some small table scripts as vpinball itself comes with [it's own scripts](https://github.com/vpinball/vpinball/tree/master/scripts).

As I understood there are other things that should be improved before we can start thinking about (a) different scripting engine(s). Eg reducing the COM related code.

**Comment by toxieainc** [comment]
Exactly. This will be a process that will continue over the years.

There are more construction sites currently in flight that need to be resolved first, step by step, then at some point i'm sure that the scripting will also be revisited as the dust settles.

**Comment by thelayeredmind** [comment]
@francisdb I had no idea, how far-reaching the scripts are, sorry for this oversight. This is indeed a harder task than it first looks. Let me know if I can be of any assistance, once, and if the team is ready to move on. I like to flex some of the polyglot muscles and this seems like a rewarding application for it.

**Comment by francisdb** [comment]
> Exactly. This will be a process that will continue over the years.
> 
> There are more construction sites currently in flight that need to be resolved first, step by step, then at some point i'm sure that the scripting will also be revisited as the dust settles.

Maybe keeping a list of construction sites would not be a bad idea?

**Comment by Lovinity** [comment]
This is something we will definitely have to consider for Visual Pinball before 2027 when the DLLs stop working (or hopefully by then VPE will be stable and we can deprecate VPX and migrate everything to VPE instead).

The potential issue I see is that Visual Basic was easy to use because it is part of the "Visual" family of languages (such as C) which VPX itself is coded. Thus it could integrate nicely. It was also easy to learn.

Other languages might not work well with C languages. VPX would have to compile them into C under the hood in order to use them. So the best option would just be to use C. But it isn't nearly as friendly of a language to use, as mentioned before, for coding game logic.

Microsoft recommended JavaScript. JS is a nice language but it's more web-oriented. I don't think it would work well for VPX. It is also very messy and prone to making bugs (thus why Typescript was invented).

While I'm personally not a big fan of the strict indentation rules, perhaps Python may be the way to go. It's widely supported. It is easy to learn and understand. And there are compilers for Python into virtually any other language... I'm sure there's one to compile it to C.

**Comment by toxieainc** [comment]
As e.g. VPX standalone uses its own/Wines VBScript interpreter, backwards compatibility will always be there in the future, so no need to worry on that part.

**Comment by racemaniac** [comment]
If you'd move to a newer language, rather than something like javascript, why not just .NET?

Then people who like the VBScript syntax, can use the more modern visual basic language supported by .NET.
For a more modern programming style, you can go for C#.
There is even F# for those who prefer functional programming :).

And with the current versions of .NET being open source, even if microsoft some day abandons them, who cares :).
